### PR TITLE
refactor common_fused_moe.py

### PR DESCRIPTION
**Pull request by Pr0Wh1teGivee**

### CODE REVIEW

**Comment by weijinqian0** [comment]
This is part of our Moe module refactoring, primarily aimed at reducing redundant code.
image
1 Refactoring Principles: Single Responsibility + Lightweight Abstraction

Issue:
· Both quantitative and non-quantitative scenarios implement select_experts, with the logic being scoring → top_k. The code duplication rate is over 90%, requiring "multiple maintenance" for modifications, which is highly prone to introducing defects. 

image
Plan:
· Refactor `select_experts` into a standalone module `experts_selector.py`, serving as a pure function for all Methods to call;
· Utilize parameters such as `e_score_correction_bias` and `scoring_func` along with lightweight branching to mask differences, avoiding the introduction of new classes and maintaining concise calls;
· Retain dual implementations of `_native_select_experts` and `_select_experts_with_fusion_ops` to support flexible switching between native PyTorch and NPU operators.

image

Value:
· One modification, three scenarios affected, eliminating scattergun changes;
· Over 80% reduction in code lines, with converged logic.

**Comment by weijinqian0** [comment]
2  Refactor dispatch/apply_mlp/combine — Separation of Concerns + Unified Entry Point

Refactoring Principle: Strategy Pattern + Open/Closed Principle
Issue: The original fused_experts_xxx function (such as fused_experts_with_mc2) tightly couples the dispatch, mlp, and combine stages, and has separate implementations for quantized and non-quantized versions, leading to "shotgun modifications." 

image


Plan:
· Phase Decoupling: Split fused_experts into three distinct phases:
a) token_dispatcher.token_dispatch(...)
b) unified_apply_mlp(...)
c) token_dispatcher.token_combine(...)
· MLP Unification: Introduce unified_apply_mlp to uniformly handle scenarios such as quantization/non-quantization and w8a8/w4a8 through parameters like with_quant, fusion, and need_trans.

image

**Comment by weijinqian0** [comment]
3 Communication Preparation/Tail Logic Encapsulation – Solving the Branching Hell
Refactoring Principle: Strategy Pattern + Interface Segregation
Problem:
· The prepare/finalize logic in AscendFusedMoE.forward is cluttered with numerous if-else branches, determining the communication mode based on FusedMoEState, leading to bloated and hard-to-maintain code. 
image


Plan:
· Introduce the FusedMoEPrepareAndFinalize strategy base class, defining standard prepare/finalize interfaces;
Each communication method (MC2/AllGather/All2All/NaiveMulticast) implements its specific logic through subclasses, such as:
FusedMoEPrepareAndFinalizeWithMC2: Handles mc2_mask alignment and TP slicing;
FusedMoEPrepareAndFinalizeWithAllGather: Implements DP AllGather + Router recalculation; 

image

**Comment by weijinqian0** [comment]
4 Object-Oriented Binding of Communication Methods - Unified Interface, Dynamic Assembly
Refactoring Principles: Template Method + Factory Method + Dependency Inversion
Issues:
- Components required for various communication methods (Dispatcher + Prepare/Finalize) are initialized separately, with non-uniform entry points, making external usage complex.
Solution:
- Introduce a base class MoECommMethod as an abstract entry point for unified communication strategies;
- During subclass initialization, dynamically bind corresponding instances through the factory methods _get_token_dispatcher() and _get_fused_moe_prepare_finalize();
- The parent class implements standard interfaces: prepare() / fused_experts() / finalize(), internally calling the bound components to shield implementation details from the outside;
- External callers only need to interact with the unified interface, without needing to be aware of underlying communication or quantization differences. 


image

Value:
· Adheres to the "Hollywood Principle," where the caller does not need to concern itself with the internal implementation;
· Complies with the "Open/Closed Principle," allowing new communication methods to be added by simply creating new subclasses without modifying any existing code;
· Clear architectural layering: Interface Layer (MoECommMethod) → Strategy Layer (Dispatcher/Prepare & Finalize) → Implementation Layer (NPU Ops).

**Comment by weijinqian0** [comment]
5 General Design

image

**Comment by weijinqian0** [comment]
approved

File: vllm_ascend/ops/moe/moe_mlp.py [hunk 1]
```diff
@@ -77,7 +76,7 @@ def quant_apply_mlp(hidden_states: torch.Tensor,
     bias1, bias2 = None, None
     _output_dtype = w2_scale.dtype
 
-    is_mc2 = get_forward_context().fused_moe_state == FusedMoEState.MC2
+    is_mc2 = get_forward_context().moe_comm_method_name == "mc2commimpl"
```
Comment (wangxiyuan) [inline vllm_ascend/ops/moe/moe_mlp.py:79]: there are a lot hardcode `xxxcommimpl`, it's good to use Enum instead.

File: vllm_ascend/ops/fused_moe.py [hunk 2]
```diff
@@ -422,63 +365,16 @@ def forward(self,
...
-                ).dp_metadata.cu_tokens_across_dp_cpu
-                hidden_states = self.naive_multicast(hidden_states,
-                                                     cu_tokens_across_dp_cpu)
-                if self.rm_router_logits:
-                    router_logits, _ = gate(hidden_states)
-                else:
-                    router_logits = self.naive_multicast(
-                        router_logits, cu_tokens_across_dp_cpu)
+        moe_comm_method_name = forward_context.moe_comm_method_name
+        forward_context.moe_comm_method = getattr(self, moe_comm_method_name)
```
Comment (wangxiyuan) [inline vllm_ascend/ops/fused_moe.py:369]: set a new property to forward_context here is not suitable, it breaks the princeple of forward_context and make the code hard for debug. While i notice that it's been done in common fused moe already. Let's refactor it in the next PR.

File: vllm_ascend/worker/model_runner_v1.py [hunk 3]
```diff
@@ -1687,19 +1688,28 @@ def _select_moe_comm_method(self, num_tokens: int) -> str:
...
-                moe_comm_method = "allgather"
+                if quant_type == "w4a8_dynamic":
+                    moe_comm_method = "alltoall"
+                else:
+                    moe_comm_method = "allgather"
+
         elif soc_version in {AscendSocVersion.A3}:
             moe_comm_method = "mc2" if num_tokens <= self.mc2_tokens_capacity else "alltoall"
         else:
             raise ValueError(f"Unsupported soc_version: {soc_version}")
```
Comment (wangxiyuan) [inline vllm_ascend/worker/model_runner_v1.py:1708]: looks that 310p is missing

File: vllm_ascend/worker/model_runner_v1.py [hunk 4]
```diff
@@ -1687,19 +1688,28 @@ def _select_moe_comm_method(self, num_tokens: int) -> str:
             str: The selected MoE communication method, either "allgather", "mc2", or "alltoall".
         """
         soc_version = get_ascend_soc_version()
+        quant_type = getattr(self.vllm_config.model_config.hf_config,
+                             'moe_quantize', None)
 
         if not self.parallel_config.enable_expert_parallel:
             moe_comm_method = "allgather"
         elif soc_version in {AscendSocVersion.A2}:
             if num_tokens <= self.mc2_tokens_capacity and self.parallel_config.world_size_across_dp >= 16:
```
Comment (wangxiyuan) [inline vllm_ascend/worker/model_runner_v1.py:1697]: previous: go into allgather if `VLLM_ENABLE_FUSED_EXPERTS_ALLGATHER_EP` enabled
now: go into allgather if `num_tokens <= self.mc2_tokens_capacity and self.parallel_config.world_size_across_dp >= 16` is False. 

is the case squal?

Comment (Pr0Wh1teGivee) [reply vllm_ascend/worker/model_runner_v1.py:1697]: Confirmed with Ant, it is ok to use _select_comm_method.

File: vllm_ascend/worker/model_runner_v1.py (hunk 3 above)
Comment (Pr0Wh1teGivee) [reply vllm_ascend/worker/model_runner_v1.py:1708]: FusedExpertsWithMoGE will be used by judging "model_type == "PanguProMoE" in the future. No need to add 310P here.

File: vllm_ascend/ops/fused_moe.py (hunk 2 above)
Comment (Pr0Wh1teGivee) [reply vllm_ascend/ops/fused_moe.py:369]: Fair point, will be refactored in my next pr.

File: vllm_ascend/ops/moe/moe_mlp.py (hunk 1 above)
Comment (Pr0Wh1teGivee) [reply vllm_ascend/ops/moe/moe_mlp.py:79]: Fair point, will be refactored in my next pr.

**Comment by wangxiyuan** [comment]
pangu is broken by qwen3-next PR. issue is here: https://github.com/vllm-project/vllm-ascend/issues/2949 We'll fix it later.
